Refactored Windows perfmon metricset: replaced method to retrieve counter paths with PdhExpandWildCardPathW, separated code by responsibility, removed unused functions

narph opened the pull request:
- Using the [PdhExpandWildCardPathW](https://docs.microsoft.com/en-us/windows/desktop/api/pdh/nf-pdh-pdhexpandwildcardpathw) func to retrieve all counter paths instead of PdhGetFormattedCounterArrayW.
Will fix https://github.com/elastic/beats/issues/10660
 -  Separated methods in different go files by responsibility (will be useful for the iis module as well)
 -  Added extra tests for query, pdh and reader 
 -  Removed unused functionality and refactored the retrieval of counter values by format

ruflin commented on metricbeat/module/windows/perfmon/_meta/docs.asciidoc:
```diff
@@ -41,7 +42,27 @@ event. In the above example, this will cause the physical_disk.write.per_sec
+*`append_instance_counter`*:: A boolean option to retrieve the full instance name with the included
+ counter, ex.
+----
+ "perfmon" : {
```
Indentation looks off here and below. Not sure if it's just Github (did not build the docs).

ruflin commented on metricbeat/module/windows/perfmon/_meta/fields.yml:
```diff
@@ -1 +1 @@
-- release: beta
+release: beta
```
I think the previous version was correct?

ruflin commented on metricbeat/module/windows/perfmon/pdh_query_windows.go:
```diff
@@ -0,0 +1,211 @@
+		return nil, err
+	}
+	return UTF16ToStringArray(expdPaths), nil
+
```
Nit: remove newline

ruflin commented on metricbeat/module/windows/perfmon/pdh_query_windows.go:
```diff
@@ -0,0 +1,211 @@
+	for path, counter := range q.counters {
+		_, value, err := PdhGetFormattedCounterValue(counter.handle, counter.format|PdhFmtNoCap100)
+		if err != nil {
+			rtn[path] = append(rtn[path], CounterValue{Err: err})
```
I was reading this line and 116 / 118 a few times. Is `rtn[path]` always set or could we run into a panic here?

jsoriano replied:
This shouldn't be a problem here, appending to a nil array works.

ruflin commented on metricbeat/module/windows/perfmon/pdh_query_windows.go:
```diff
@@ -0,0 +1,211 @@
+}
+
+// Error returns a more explicit error message.
+func (e PdhErrno) Error() string {
```
Do we have some tests for this method somehwere? Pretty hard to follow what the end outcome is without a test.

ruflin replied:
BTW I wonder if this method belongs here or rather in the phd_windows file.

narph replied:
indeed, there is already a test in pdh_windows_test file, and I have moved this func to the pdh_windows file

narph commented:
@jsoriano , I have removed the 3rd change from this PR (below):

 - Added `perfmon.append_instance_counter: true` config option in order to retrieve the full instance name (with counter), based on issue https://github.com/elastic/beats/issues/9835 and PR https://github.com/elastic/beats/pull/9847
 If enabled, this will change:
```
      "windows" : {
            "perfmon" : {
              "processor" : {
                "name" : "RuntimeBroker",
                "time" : {
                  "total" : {
                    "pct" : 0
                  }
                }
              }
            }
          },
```
to:
```
"windows" : {
            "perfmon" : {
              "processor" : {
                "name" : "RuntimeBroker#4",
                "time" : {
                  "total" : {
                    "pct" : 0
                  }
                }
              }
            }
          },
```

as I will try to handle the configuration changes in a different PR.
What has been done so far here:
- removed unused functionality
- separate responsibility in different files (added reader and pdh query)
- fixed some wildcard related issues by retrieving the counter with [pdhexpandwildcardpathw](https://docs.microsoft.com/en-us/windows/desktop/api/pdh/nf-pdh-pdhexpandwildcardpathw).
- added tests for reader, query and pdh.

jsoriano commented on metricbeat/module/windows/perfmon/pdh_windows_test.go:
```diff
@@ -23,9 +23,86 @@ import (
+		counters = append(counters, counterHandle)
+	}
+	err = PdhCollectQueryData(queryHandle)
+	err = PdhCollectQueryData(queryHandle)
```
Same here, is this needed twice here? Add a comment if it is.

narph replied:
same as the above

jsoriano commented on metricbeat/module/windows/perfmon/zpdh_windows.go:
```diff
@@ -56,16 +56,12 @@ func errnoErr(e syscall.Errno) error {
-	procPdhGetRawCounterArray           = modpdh.NewProc("PdhGetRawCounterArray")
-	procPdhCalculateCounterFromRawValue = modpdh.NewProc("PdhCalculateCounterFromRawValue")
-	procPdhFormatFromRawValue           = modpdh.NewProc("PdhFormatFromRawValue")
-	procPdhCloseQuery                   = modpdh.NewProc("PdhCloseQuery")
```
Nice cleanup here :slightly_smiling_face:

jsoriano commented on metricbeat/module/windows/perfmon/reader.go:
```diff
@@ -0,0 +1,172 @@
+			}
+		}
+		if childQueries == nil || len(childQueries) == 0 {
+			return nil, errors.Wrapf(err, `failed to expand counter (query="%v")`, counter.Query)
```
`query.Close()` here too, or use this [`cleanup`](https://github.com/elastic/beats/pull/12134) helper.

jsoriano commented on metricbeat/module/windows/perfmon/reader.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+// CloseQuery will close the PDH query.
+func (r *Reader) CloseQuery() error {
```
Just call this method `Close`, so it doesn't expose implementation details and can be used if more things need to be closed in a Reader in the future.
```suggestion
func (r *Reader) Close() error {
```

jsoriano commented on metricbeat/module/windows/perfmon/perfmon.go:
```diff
@@ -102,3 +100,12 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) {
 }
+
+// Close will be called when metricbeat is stopped, should close the query.
+func (m *MetricSet) Close() error {
```
Thanks for implementing this closer :+1: 

It could be also written in one line as:
```
return errors.Wrap(m.reader.Close(), "failed to close perfmon reader"))
```
But I don't have a strong preference, so as you prefer.

narph commented on metricbeat/module/windows/perfmon/pdh_query_windows.go:
```diff
@@ -0,0 +1,187 @@
+			rtn[path] = append(rtn[path], CounterValue{Measurement: *(*float64)(unsafe.Pointer(&value.LongValue)), Instance: counter.instanceName})
+		case PdhFmtLarge:
+			rtn[path] = append(rtn[path], CounterValue{Measurement: *(*int64)(unsafe.Pointer(&value.LongValue)), Instance: counter.instanceName})
+
```
@jsoriano , the recent commit should implement all format options (long, float, large), we might need to update current documentation:
> Format of the measurement value. The value can be either float or long. The default is float.

jsoriano replied:
If we make double the default I think we should have an explicit `case "double", "":` so only `double` or the empty string are used for doubles, and any other unknown value should provoke an error, mentioning the expected values, as this case is probably caused by a typo in the config.
And then `getCounterValue` should only have cases for valid formats, and report an error on unexpected values.

narph commented on metricbeat/module/windows/perfmon/pdh_query_windows_test.go:
```diff
@@ -0,0 +1,87 @@
+		t.Fatal(err)
+	}
+	err = q.CollectData()
+	err = q.CollectData()
```
added a comment explaining the calls (thanks for pointing that out):
//Some counters, such as rate counters, require two counter values in order to compute a displayable value. In this case we must call PdhCollectQueryData twice before calling PdhGetFormattedCounterValue.
	// For more information, see Collecting Performance Data (https://docs.microsoft.com/en-us/windows/desktop/PerfCtrs/collecting-performance-data).

prakash1243 commented:
@narph - I have checked out your pull request and built meticbeat package locally to collect Windows Perfmon Counters on a Windows-7 32 bit server, I have added a wild card query to collect Private Bytes across all the processes running, The output seems to be different now:

Configuration:
`  - instance_label: process_private_bytes
    measurement_label: process.private.bytes
    query: '\Process(*)\Private Bytes'

Output:
` "metricset": {
      "name": "perfmon",
      "module": "windows",
      "rtt": 32001
    },
    "event": {
      "dataset": "windows.perfmon",
      "duration": 32001800
    },
    "windows": {
      "perfmon": {
        "process_private_bytes": "*",
        "process": {
          "private": {
            "bytes": 0
          }
        }
      }
    }
  }`

It doesn't collect the Private Bytes for all the processes, instead it had sent only one document with the process name as '*' . I have also added 'perfmon.append_instance_counter: true' to my windows configuration. 
I have been waiting for my issue to get fixed #10660, got to know it would be fixed with this PR, so just did a test and wanted to let you know my findings. Thanks !

narph commented:
@prakash1243, I am getting different results here when testing locally with `- instance_label: process_private_bytes measurement_label: process.private.bytes query: '\Process(*)\Private Bytes`
```
},
          "metricset" : {
            "name" : "perfmon"
          },
          "service" : {
            "type" : "windows"
          },
          "windows" : {
            "perfmon" : {
              "process" : {
                "private" : {
                  "bytes" : 2564096.0
                }
              },
              "process_private_bytes" : "winpty-agent"
            }
          },
          "ecs" : {
            "version" : "1.0.0"
          },
```
Nonetheless, I have added an integration test `TestWildcardQueryNoInstanceName` in `pdh_integration_windows_test.go` which should pass, mind giving that one a go?
Can you provide me with the entire windows module config if you are still dealing with this issue.

Regarding `perfmon.append_instance_counter` , in later conversations (https://github.com/elastic/beats/pull/12212#issuecomment-495273843) I am pointing out that I would leave this feature out and integrate it in a new PR.

prakash1243 commented:
Hi @narph - Here is module config:

`- module: windows
  metricsets: [perfmon]
  period: 5s
  perfmon.counters:
  - instance_label: process_private_bytes
    measurement_label: process.private.bytes
    query: '\Process(*)\Private Bytes'
`
Will I be able to get a snapshot build with the latest code changes you have  in your PR for the Windows 32 bit metricbeat.exe? I can just test and let you know. 
I doubt if am doing something wrong while building the package; since am new to the Go lang. :) 
Please let me know. Thanks !

prakash1243 commented:
Hi @narph - 
Am getting this error, when I try to build the metricbeat package :(
`C:\Users\Admin\go\src\github.com\elastic\beats\metricbeat>go build
# github.com/elastic/beats/vendor/github.com/DataDog/zstd
In file included from ./zstd.h:43:0,
                 from ..\vendor\github.com\DataDog\zstd\zstd.go:4:
C:/Users/Admin/Desktop/gcc-5.1.0-tdm-1-core/lib/gcc/mingw32/5.1.0/include/stdint
.h:9:26: fatal error: stdint.h: No such file or directory
compilation terminated.`

Would you mind sharing the snapshot build, which could be a huge help for me to test and let you know my findings. Thanks !

narph commented:
@prakash1243 , I believe you can download a snapshot from one of the latest CI builds here https://beats-ci.elastic.co/job/elastic+beats+master+multijob-package-linux/663/gcsObjects/

prakash1243 commented:
> @prakash1243 , I believe you can download a snapshot from one of the latest CI builds here https://beats-ci.elastic.co/job/elastic+beats+master+multijob-package-linux/663/gcsObjects/
@narph - does this snapshots builds include the changes you had on this PR?

ruflin commented on metricbeat/module/windows/perfmon/perfmon.go:
```diff
@@ -63,22 +63,20 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 		form := strings.ToLower(value.Format)
 		switch form {
-		case "":
+		case "", "float":
```
Was this a bug before?

jsoriano commented on metricbeat/module/windows/perfmon/pdh_query_windows.go:
```diff
@@ -0,0 +1,200 @@
+		return PdhFmtLong
+	case "large":
+		return PdhFmtLarge
+	default:
```
Ok, I see that we check for proper values when reading the config, but I'd still feel better if we checked for valid values here and in `getCounterValue` :grimacing:

narph replied:
I have added an extra check when retrieving the counter values, thanks for the feedback.

andrewkroh commented on metricbeat/module/windows/perfmon/pdh_query_windows.go:
```diff
@@ -0,0 +1,208 @@
+// AddCounter adds the specified counter to the query.
+func (q *Query) AddCounter(counterPath string, counter CounterConfig, wildcard bool) error {
+	if _, found := q.counters[counterPath]; found {
+		return errors.Errorf("Counter %s has been already added", counterPath)
```
Error messages should not be capitalized. https://github.com/golang/go/wiki/CodeReviewComments#error-strings

andrewkroh commented on metricbeat/module/windows/perfmon/pdh_query_windows.go:
```diff
@@ -0,0 +1,208 @@
+	}
+}
+
+// UTF16ToStringArray converts list of Windows API NULL terminated strings  to go string array.
```
```suggestion
// UTF16ToStringArray converts list of Windows API NULL terminated strings to a Go string array.
```

narph commented:
jenkins, test this
